deeppavlov/models/kbqa/query_generator.py

```diff
@@ -70,164 +76,317 @@ def __call__(self, question_batch: List[str],
-                    entity_tags_batch, answer_types_batch):
+                 probas_batch: List[List[float]],
+                 answer_types_batch: List[Set[str]] = None,
+                 entities_to_link_batch: List[List[int]] = None) -> Tuple[List[Any], List[Any]]:
+
+        candidate_outputs_batch, template_answers_batch = [], []
+        if not answer_types_batch or answer_types_batch[0] is None:
+            answer_types_batch = [[] for _ in question_batch]
+        if not entities_to_link_batch or entities_to_link_batch[0] is None:
+            entities_to_link_batch = [[1 for _ in substr_list] for substr_list in entities_from_ner_batch]
+        log.debug(f"kbqa inputs {question_batch} {question_san_batch} template_type_batch: {template_type_batch} --- "
+                  f"entities_from_ner: {entities_from_ner_batch} --- types_from_ner: {types_from_ner_batch} --- "
+                  f"entity_tags_batch: {entity_tags_batch} --- answer_types_batch: "
+                  f"{[list(elem)[:3] for elem in answer_types_batch]}")
+        for question, question_sanitized, template_type, entities_from_ner, types_from_ner, entity_tags_list, \
+            probas, entities_to_link, answer_types in zip(question_batch, question_san_batch, template_type_batch,
+                                                          entities_from_ner_batch, types_from_ner_batch,
+                                                          entity_tags_batch, probas_batch, entities_to_link_batch,
+                                                          answer_types_batch):
             if template_type == "-1":
                 template_type = "7"
-            candidate_outputs, template_answer, templates_nums = \
+            candidate_outputs, template_answer = \
                 self.find_candidate_answers(question, question_sanitized, template_type, entities_from_ner,
-                                            entity_tags_list, answer_types)
+                                            types_from_ner, entity_tags_list, probas, entities_to_link, answer_types)
             candidate_outputs_batch.append(candidate_outputs)
             template_answers_batch.append(template_answer)
-            templates_nums_batch.append(templates_nums)
 
-        answers = self.rel_ranker(question_batch, candidate_outputs_batch, entities_from_ner_batch,
-                                  template_answers_batch)
-        log.debug(f"(__call__)answers: {answers}")
-        if not answers:
-            answers = ["Not Found" for _ in question_batch]
-        return answers
+        if self.return_answers:
+            answers = self.rel_ranker(question_batch, template_type_batch, candidate_outputs_batch,
+                                      entities_from_ner_batch, template_answers_batch)
+            log.debug(f"(__call__)answers: {answers}")
+            if not answers:
+                answers = ["Not Found" for _ in question_batch]
+            return answers
```

Comment:
There are two options: if we want to perform rel ranking in query_generator component, then self.return_answers is true and we output answers, otherwise we output candidate_outputs_batch, template_answers_batch and perform ranking in a separate component of the chainer.